### PR TITLE
Fix FE_Face elements for use in hp

### PR DESCRIPTION
**Pull request by simfeld**
Implemented the functions `hp_vertex_dof_identities,` `hp_line_dof_identities` and `hp_quad_dof_identities` for `FE_FaceQ` and `FE_FaceP.` These functions are required when using FE_Face elements in an hp finite element context. Since these elements are discontinuous, there is nothing to be returned.

This fixes the issue described [here](https://groups.google.com/d/msg/dealii/RF8ZZLXrHA4/sloD1vbaCQAJ).

### CODE REVIEW

**Comment by bangerth** [comment]
First, welcome to submitting your first patch! I hope this is only the first in a long list!

I will admit that I'm a bit confused by why the functions all return empty sets of constraints. The documentation you add says that the element is discontinuous, but that isn't quite true, is it? If you had two neighboring elements that differ in polynomial degree, wouldn't you want the functions on the two sides of the face to be the same? My understanding is that the element is only discontinuous where faces come together (at vertices in 2d, and edges in 3d) but that you would still want the two sides of the same face to match. Or am I wrong?

(I know that the answer in the forum suggested to copy these functions from `FE_DGQ`, but I'm not in fact sure that's actually the correct approach.)

**Comment by bangerth** [comment]
(That said, it is correct that the constraints are empty for (i) vertices in 2d, and (ii) vertices and lines in 3d.)

**Comment by kronbichler** [comment]
I agree with @bangerth here: The elements put all degrees of freedom into the face, which is shared between the two cells adjacent to that. Thus, there must be some constraints on those indices, but as @bangerth points out there will be none on the entities below the faces.

**Comment by masterleinad** [comment]
I suggest that we fix up the implementations for `FE_FaceQ` and drop the ones for `FE_FaceP` unless anyone sees how the `hp_*_dof_identities` could be implemented in case of different polynomial degrees on a shared face.

**Comment by simfeld** [comment]
I see your point @bangerth and @kronbichler . I will try to fix this as soon as possible (right now i am on holiday so it will take some time).

File: source/fe/fe_face.cc [hunk 1]
```diff
@@ -263,6 +263,158 @@ FE_FaceQ<dim,spacedim>::hp_constraints_are_implemented () const
...
+
+
+
+template <int dim, int spacedim>
+std::vector<std::pair<unsigned int, unsigned int> >
+FE_FaceQ<dim, spacedim>::
+hp_line_dof_identities (const FiniteElement<dim, spacedim> &fe_other) const
+{
+  // this element is continuous only for the highest dimensional bounding object
+  if (dim !=2)
```
Comment (bangerth) [inline source/fe/fe_face.cc]: can you write this as `if (dim > 2)` to make clear the intent?

File: source/fe/fe_face.cc [hunk 2]
```diff
@@ -263,6 +263,158 @@ FE_FaceQ<dim,spacedim>::hp_constraints_are_implemented () const
...
+
+
+
+template <int dim, int spacedim>
+std::vector<std::pair<unsigned int, unsigned int> >
+FE_FaceQ<dim, spacedim>::
+hp_quad_dof_identities (const FiniteElement<dim, spacedim> &fe_other) const
+{
+  // this element is continuous only for the highest dimensional bounding object
+  if (dim !=3)
```
Comment (bangerth) [inline source/fe/fe_face.cc]: Same here -- maybe write it as `if (dim > 3)`.

I *think* that this function can not actually be called if `dim==1` or `dim==2`. But we can verify that by putting an `Assert (dim >= 3, ExcInternalError())` at the top. This then also explains that the `else` part below here really just means `dim==3`.

**Comment by bangerth** [comment]
Oh, well, we let you wait for such a long time with this PR that I'll just merge and do the changes myself.

Many thanks for working on this and getting it into good shape!

**Pull request merged**